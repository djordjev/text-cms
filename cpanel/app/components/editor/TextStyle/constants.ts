export const TEXT_STYLES = [
  { id: 'heading_1', name: 'Heading 1' },
  { id: 'heading_2', name: 'Heading 2' },
  { id: 'heading_3', name: 'Heading 3' },
  { id: 'heading_4', name: 'Heading 4' },
  { id: 'heading_5', name: 'Heading 5' },
  { id: 'heading_7', name: 'Heading 6' },
  { id: 'paragraph', name: 'Paragraph' }
];
