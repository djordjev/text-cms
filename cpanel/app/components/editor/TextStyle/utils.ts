import { BaseEditor, Editor, Element } from 'slate';
import { ReactEditor } from 'slate-react';

import { CustomElement } from '~/components/editor/VariationEditor/types';

const getSelectedStyle = (editor: BaseEditor & ReactEditor) => {
  const { selection } = editor;

  if (!selection) return;

  const [match] = Array.from(
    Editor.nodes<CustomElement>(editor, {
      at: selection,
      match: (n) =>
        !Editor.isEditor(n) &&
        Element.isElement(n) &&
        (n.type === 'paragraph' || n.type === 'heading')
    })
  );

  const [nodeType] = match;

  return nodeType;
};

export { getSelectedStyle };
