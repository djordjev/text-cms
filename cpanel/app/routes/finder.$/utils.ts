import { CreateParam } from '~/types';

const getCreateOpenModal = (params: URLSearchParams): CreateParam | null => {
  const param = params.get('action');

  if (param === 'new-file' || param === 'new-folder') return param;

  return null;
};

export { getCreateOpenModal };
