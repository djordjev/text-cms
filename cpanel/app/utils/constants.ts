export const COLORS = {
  BLACK: '#000',
  PRIMARY: '#003EB3'
};

export const BUTTON_ACTION = '_action';
