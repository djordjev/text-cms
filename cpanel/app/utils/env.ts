const readEnvVariable = <T = string>(name: string) => {
  const vars = process.env;

  const val = vars[name];
  if (!val) throw new Error(`Variable ${name} not supplied`);

  return val as T;
};

export { readEnvVariable };
