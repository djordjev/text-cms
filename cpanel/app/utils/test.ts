import { Descendant } from 'slate';

export const TEST_EDITOR: Descendant[] = [
  {
    type: 'heading',
    children: [
      {
        text: 'HEADING'
      }
    ],
    level: 3,
    align: 'center'
  },
  {
    type: 'paragraph',
    children: [
      {
        text: 'A '
      },
      {
        text: 'line',
        italic: true
      },
      {
        text: ' of '
      },
      {
        text: 'text',
        color: '#e41111',
        bold: true
      },
      {
        text: ' in a '
      },
      {
        text: 'paragraph',
        bold: true
      },
      {
        text: '.'
      }
    ]
  }
];

export const buildRequest = (data: Record<string, string>, method = 'POST') => {
  const form = new URLSearchParams(data);

  return new Request('https://www.test.local', { body: form, method });
};
