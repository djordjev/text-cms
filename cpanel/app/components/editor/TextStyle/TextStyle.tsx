import React, { FC } from 'react';
import { useSlate } from 'slate-react';

import { getSelectedStyleOption } from '~/components/editor/TextStyle/helpers';
import { CustomElement } from '~/components/editor/VariationEditor/types';

import { TEXT_STYLES } from './constants';
import { getSelectedStyle } from './utils';

export interface TextStyleProps {
  onChange: (selected: Partial<CustomElement>) => void;
}

const TextStyle: FC<TextStyleProps> = (props) => {
  const { onChange } = props;

  // Hooks
  const editor = useSlate();

  // Setup
  const selectedStyle = getSelectedStyle(editor);
  const selected = getSelectedStyleOption(selectedStyle);

  // Handler
  const onChangeHandler = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const id = e.currentTarget.value;

    if (id === 'paragraph') {
      onChange({ type: id });
      return;
    }

    const level = id.split('_')[1] ?? '3';
    onChange({ level: Number.parseInt(level, 10), type: 'heading' });
  };

  return (
    <select
      className={'u-bg-gray-1'}
      onChange={onChangeHandler}
      value={selected}
    >
      {TEXT_STYLES.map((style) => (
        <option key={style.id} value={style.id}>
          {style.name}
        </option>
      ))}
    </select>
  );
};

export { TextStyle };
