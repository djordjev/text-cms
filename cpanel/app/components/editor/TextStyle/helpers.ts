import { BaseEditor, Editor, Element } from 'slate';
import { ReactEditor } from 'slate-react';

import { TEXT_STYLES } from '~/components/editor/TextStyle/constants';
import { CustomElement } from '~/types/editor';

const getSelectedStyleOption = (selected?: CustomElement) => {
  if (!selected) return TEXT_STYLES[TEXT_STYLES.length - 1].id;

  const { type } = selected;

  if (type === 'heading') {
    const { level = 3 } = selected;
    return `${type}_${level}`;
  }

  return type;
};

const getSelectedStyle = (editor: BaseEditor & ReactEditor) => {
  const { selection } = editor;

  if (!selection) return;

  const [match] = Array.from(
    Editor.nodes<CustomElement>(editor, {
      at: selection,
      match: (n) =>
        !Editor.isEditor(n) &&
        Element.isElement(n) &&
        (n.type === 'paragraph' || n.type === 'heading')
    })
  );

  const [nodeType] = match;

  return nodeType;
};
export { getSelectedStyle, getSelectedStyleOption };
