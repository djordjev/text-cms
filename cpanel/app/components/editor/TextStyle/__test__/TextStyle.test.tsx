import { createRemixStub } from '@remix-run/testing';
import { fireEvent, render, screen } from '@testing-library/react';
import { createEditor } from 'slate';
import { Slate, withReact } from 'slate-react';

import { TEST_EDITOR } from '~/utils/test';

import { TextStyle, TextStyleProps } from '../TextStyle';

test('TextStyle', () => {
  const onChange = vi.fn();
  const editor = withReact(createEditor());

  const props: TextStyleProps = { onChange };

  const createComponent = () => {
    const RemixStub = createRemixStub([
      {
        path: '/',
        Component: () => (
          <Slate editor={editor} initialValue={TEST_EDITOR}>
            <TextStyle {...props} />
          </Slate>
        )
      }
    ]);

    return <RemixStub />;
  };

  const { container } = render(createComponent());

  const options = [
    'Heading 1',
    'Heading 2',
    'Heading 3',
    'Heading 4',
    'Heading 5',
    'Heading 6',
    'Paragraph'
  ];

  options.forEach((opt) => expect(screen.getByText(opt)).toBeInTheDocument());

  expect(container.childNodes).toMatchSnapshot();

  const select = container.firstChild!;

  // Select paragraph
  fireEvent.change(select, { currentTarget: { value: 'paragraph' } });

  expect(onChange).toHaveBeenCalledOnce();
  expect(onChange).toHaveBeenCalledWith({ type: 'paragraph' });
});
