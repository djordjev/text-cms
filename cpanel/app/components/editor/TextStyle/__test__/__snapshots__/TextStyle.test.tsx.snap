// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextStyle 1`] = `
NodeList [
  <select
    class="u-select-sm u-select-bordered"
  >
    <option
      value="heading_1"
    >
      Heading 1
    </option>
    <option
      value="heading_2"
    >
      Heading 2
    </option>
    <option
      value="heading_3"
    >
      Heading 3
    </option>
    <option
      value="heading_4"
    >
      Heading 4
    </option>
    <option
      value="heading_5"
    >
      Heading 5
    </option>
    <option
      value="heading_7"
    >
      Heading 6
    </option>
    <option
      value="paragraph"
    >
      Paragraph
    </option>
  </select>,
]
`;
